Adobe esign tests

shraddhacloudelements opened the pull request:
# Highlights
--Adobe ESign Element (develop-CE-3676)

--Churros integration tests for agreementAssetEvents, geturls, groups, libraryDocuments, transientDocuments, users, widgets and workflows.

# Notes
--Three of the scripts in LibraryDocuments are not working, probably due to time-out issue. The test suite might be running too fast. All these APIs works fine with Elements Builder. I have commented them in 'libraryDocuments.js'.

--One of the script in Widgets, 'GET /widgets' has stopped working probably because around 500 widgets are created in the APP and it times-out in loading all the widgets. It works fine with Elements Builder, but takes some time to load. It used to work fine earlier. Also, there is no way to delete those widgets, neither manually in the APP nor through API. I have commented the script in 'widgets.js'.

Could you please take a look in the above issues and suggest me how to deal with them?

# Results
####  agreementAssetEvents
    ✓ should allow GET for /hubs/esignature/agreementAssetEvents (16567ms)

####  geturls
    ✓ should allow GET for /hubs/esignature/geturls (413ms)

#### groups
    ✓ should allow CRUDS for /hubs/esignature/groups (4014ms)
    ✓ should allow GET for /hubs/esignature/groups/{groupId}/users (2684ms)

####  libraryDocuments
    ✓ should allow GET for /hubs/esignature/libraryDocuments (5479ms)
    ✓ should allow POST for /hubs/esignature/libraryDocuments (1543ms)
    ✓ should allow GET for /hubs/esignature/libraryDocuments{libraryDocumentId} (2898ms)
    ✓ should allow GET for /hubs/esignature/libraryDocuments{libraryDocumentId}/auditTrail (6140ms)

####  transientDocuments
    ✓ should allow POST for /transientDocuments (775ms)

####  users
    ✓ should allow GET for /hubs/esignature/users (1783ms)
    ✓ should allow POST for /hubs/esignature/users (2373ms)
    ✓ should allow GET for /hubs/esignature/users/{userId} (2247ms)
    ✓ should allow PUT for /hubs/esignature/users/{userId} (3790ms)

####  widgets
    ✓ should allow POST for /hubs/esignature/widgets (7055ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId} (6332ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/agreements (5955ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/auditTrail (6881ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/combinedDocument (5719ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/documents (8282ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/documents/{documentId} (9286ms)
    ✓ should allow GET /hubs/esignature/widgets/{widgetId}/formData (7315ms)
    ✓ should allow PATCH for /hubs/esignature/widgets/{widgetId}/personalize (9804ms)
    ✓ should allow PATCH for /hubs/esignature/widgets/{widgetId}/status (29692ms)

####  workflows
    ✓ should allow GET for /hubs/esignature/workflows (469ms)
    ✓ should allow GET for /hubs/esignature/workflows/{workflowId} (1009ms)
    ✓ should allow POST for /hubs/esignature/workflows/{workflowId}/agreements (2035ms)
    
  26 passing (3m)

jjwyse commented on src/test/elements/adobe-esign/agreementAssetEvents.js:
```diff
@@ -0,0 +1,26 @@
+const util = require('util');
+
+suite.forElement('esignature', 'agreementAssetEvents', null, (test) => {
+  let currentDate;	
```
no need to do `let currentDate;` here, since you're assigning it two lines down.  can reduce the lines of code and clean these up by just doing `let currentDate = date.toISOString().substring(0, 19);`, etc.

also, is there a better way to do these dates as opposed to the magic numbers (19, 10)?  if not, can you just assign the magic numbers as constants, so people coming into this code know what those numbers represent?

shraddhacloudelements replied:
Done.

jjwyse commented on src/test/elements/adobe-esign/getUrls.js:
```diff
@@ -0,0 +1,13 @@
+const expect = chakram.expect;
+
+suite.forElement('esignature', 'geturls', null, (test) => {
+  it('should allow GET for ' + '/hubs/esignature/geturls', () => {
```
no need for string concatenation here

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/getUrls.js:
```diff
@@ -0,0 +1,13 @@
+
+suite.forElement('esignature', 'geturls', null, (test) => {
+  it('should allow GET for ' + '/hubs/esignature/geturls', () => {
+    return cloud.get(test.api)
```
also, instead of defining your own test here, you could just use `test.should.return200OnGet`

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/groups.js:
```diff
@@ -0,0 +1,24 @@
+
+suite.forElement('esignature', 'groups', group(), (test) => {
+  test.should.supportCruds(chakram.put);
+  it('should allow GET for ' + test.api + '/' + '{groupId}/users', () => {
```
can use the ES6 template literals, to clean up string concatenation here by using the ticks: \`should allow GET for ${test.api}/{groupId}/users\`

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/libraryDocuments.js:
```diff
@@ -0,0 +1,84 @@
+  let libraryDocumentId;
+  let transientDocumentId;
+  let documentId;
+  it('should allow GET for ' + test.api, () => {
```
can use `tests.should.return200OnGet` here

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/transientDocuments.js:
```diff
@@ -0,0 +1,13 @@
+
+suite.forElement('esignature', 'transientDocuments', null, (test) => {
+  it('should allow POST for /transientDocuments', () => {
+    return cloud.postFile(test.api, __dirname + '/assets/attach.txt')
```
if no validation CB is passed in, then `cloud.METHOD` will always validate that a `200` is returned, so no need to do it yourself here too.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/users.js:
```diff
@@ -0,0 +1,52 @@
+
+suite.forElement('esignature', 'users', null, (test) => {
+  let userId;
+  it('should allow GET for ' + test.api, () => {
```
can use `test.should.return200OnGet`

shraddhacloudelements replied:
Done

shraddhacloudelements replied:
Created `test.should.supportCrs` instead.

jjwyse commented on src/test/elements/adobe-esign/users.js:
```diff
@@ -0,0 +1,52 @@
+    return cloud.get(test.api)
+      .then(r => expect(r).to.have.statusCode(200))
+  });
+  it('should allow POST for ' + test.api, () => {
```
can use `test.should.return200OnPost`

shraddhacloudelements replied:
did this: `test.withJson(createUsers()).should.return200OnPost();`

shraddhacloudelements replied:
Created `test.should.supportCrs` instead.

jjwyse commented on src/test/elements/adobe-esign/users.js:
```diff
@@ -0,0 +1,52 @@
+    return cloud.post(test.api, createUsers())
+      .then(r => expect(r).to.have.statusCode(200))
+  });
+  it('should allow GET for ' + test.api + '/{userId}', () => {
```
might be better to create a new `test.should.supportCr` in the `test` object so we can re-use this throughout.

shraddhacloudelements replied:
Created `test.should.supportCrs` instead. This reduced few more lines in the code.

jjwyse commented on src/test/elements/adobe-esign/users.js:
```diff
@@ -0,0 +1,52 @@
+      .then(r => expect(r).to.have.statusCode(200))
+  });
+  it('should allow PUT for ' + test.api + '/{userId}', () => {
+    return cloud.get(test.api)
```
here, we're assuming there are going to be valid users in the adobe esign instance, is that a safe assumption?  i just ask as it's important that these tests do not report false negatives since they will be run as part of every pull request and if they fail, people will not be able to merge in code.

shraddhacloudelements replied:
We can do a POST /users with the same user which are 'INACTIVE' any number of time. It does not give any error and it returns the same userID every time. While if the user is ACTIVE, and we try to do a POST for the same user with same details, it gives an error. But, in our case, we do not have to worry about the ACTIVE users. A user can only be made ACTIVE when the user accepts the link and log-in to the account which is being sent in the 'email' on doing POST. Moreover, there is a license limit of just 1 member, which Greg has already used. No more user can be made ACTIVE in our developers account. So, a POST /users will always happen and if there is even a duplicate user, then also it would run. Am I able to make it clear? Please let me know if more inputs are required in this case.

jjwyse commented on src/test/elements/adobe-esign/workflows.js:
```diff
@@ -0,0 +1,53 @@
+    return cloud.get(test.api)
+      .then(r => expect(r).to.have.statusCode(200))
+  });
+  it('should allow GET for ' + test.api + '/{workflowId}', () => {
```
can you use `test.should.supportSr` for this ? (search, retrieve)

shraddhacloudelements replied:
It is difficult as to do a POST /workflows/{workflowId}/agreements, firstly we need to do POST /transientDocuments, fetch the 'transientDocumentId', substitute it in the JSON for POST, also GET the 'workflowId' from the GET /workflows and then do the POST.

shraddhacloudelements commented:
--Three of the scripts in LibraryDocuments are not working, probably due to time-out issue. The test suite might be running too fast. All these APIs works fine with Elements Builder. I have commented them in 'libraryDocuments.js'.

--One of the script in Widgets, 'GET /widgets' has stopped working probably because around 600 widgets are created in the APP and it times-out in loading all the widgets. It worked fine with Elements Builder, earlier, but it is breaking now there as well due to the same time-out issue. Also, there is no way to delete those widgets through APIs. I have commented the script in 'widgets.js'. I am in talks with the support team of Adobe Esign. Let's see if he suggests any alternate to this.

Could you please take a look at the above issues and suggest me how to deal with them?

jjwyse commented on src/test/elements/adobe-esign/agreementAssetEvents.js:
```diff
@@ -0,0 +1,27 @@
+  it('should allow GET for ' + test.api, () => {
+    return cloud.withOptions({
+      qs: {
+        where: util.format("startDate='"+startDate+"' and endDate='"+currentDate+"'")
```
i suggest using the ES6 template literals here as opposed to `util.format`.  it's easier to read and, moving forward, i've been removing all of the `util.format`s to use the template literals.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/users.js:
```diff
@@ -0,0 +1,38 @@
+});
+
+suite.forElement('esignature', 'users', null, (test) => {
+  let userId;
```
this `userId` can go inside of the `it` scope as opposed to the global scope of all tests for `users`.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/widgets.js:
```diff
@@ -0,0 +1,110 @@
+suite.forElement('esignature', 'widgets', null, (test) => {
+  let widgetId;
+  let transientDocumentId;
+  let documentId;
```
can you define these vars inside of the tests they pertain to as opposed to defining them at the global scope.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/widgets.js:
```diff
@@ -0,0 +1,110 @@
+      .then(r => cloud.patch(test.api + '/' + widgetId + '/personalize', updateWidgetPersonalize()))
+  });
+  it('should allow PATCH for ' + test.api + '/{widgetId}/status', () => {
+    return cloud.postFile('/hubs/esignature/transientDocuments', __dirname + '/assets/attach.txt')
```
since you're doing the same 3 or 4 lines here in each test, can you pull these out into a function and use that function.  that way we're not duplicating code and it will be much easier to maintain moving forward.

jjwyse commented on src/test/elements/adobe-esign/workflows.js:
```diff
@@ -0,0 +1,52 @@
+
+suite.forElement('esignature', 'workflows', null, (test) => {
+  let workflowId;
+  let transientDocumentId;
```
can you put these variables inside of the tests they pertain to as opposed to at the global scope here.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/workflows.js:
```diff
@@ -0,0 +1,52 @@
+      .then(r => expect(r).to.have.statusCode(200))
+  });*/
+  test.should.return200OnGet();
+  it('should allow GET for ' + test.api + '/{workflowId}', () => {
```
i believe this can use the `test.should.supportSr`, right?

shraddhacloudelements replied:
Done

jjwyse commented:
per our slack conversation: if you can write an issue in `churros` on why the tests are running too fast and what needs to be supported in `churros` to support it, that'd be great.  also, we'll need the cleanup selenium JS script to go about deleting any resources in the UI that may be necessary in order for these tests to run.

shraddhacloudelements commented:
## Notes

- Left the POSTs in the resources as commented in the code, as we do not have to implement it to avoid time-out issues in the GET /resource.
- Please take a look into the suite.js and cloud.js files. I have added a test for Crs in there.
- Three of the scripts in LibraryDocuments which were not working earlier, are working now. But, they fail sometime with the following error. I have created a bug for it: 
```
Uncaught Error: connection refused: localtunnel.me:43631 (check your firewall settings)
      at Socket.<anonymous> (node_modules/localtunnel/lib/TunnelCluster.js:47:32)
      at emitErrorNT (net.js:1246:8)```

jjwyse commented on src/test/elements/adobe-esign/agreementAssetEvents.js:
```diff
@@ -0,0 +1,26 @@
+  let currentDate = date.toISOString().substring(startIndex, endIndex);
+  startDate = date.setDate(date.getDate() - 10);
+  var oldDate = new Date(startDate);
+  var startDate = oldDate.toISOString().substring(startIndex, endIndex);
```
please use `let` before declaring `startDate`.  also, please be consistent and use `let` instead of `var` for these variables.

shraddhacloudelements replied:
Done

jjwyse commented on src/test/elements/adobe-esign/getUrls.js:
```diff
@@ -0,0 +1,10 @@
+const cloud = require('core/cloud');
+const expect = chakram.expect;
+
+suite.forElement('esignature', 'geturls', null, (test) => {
```
this isn't necessarily a `churros` thing, but a question nonetheless: since we're trying to make our APIs RESTful, it seems that `geturls` isn't a very good resource name here?  can we be more RESTful with our resource names in CE, even if the endpoint isn't RESTful?

shraddhacloudelements replied:
Done. Created a bug for this with CE-4170, fixed it and sent it for code review.

jjwyse commented on src/test/elements/adobe-esign/libraryDocuments.js:
```diff
@@ -0,0 +1,68 @@
+
+  test.should.return200OnGet();
+
+  it(`should allow GET for ${test.api}/{libraryDocumentId}`, () => {
```
for all of these `GET` requests, instead of duplicating code, can you use the existing:
```
test.withApi('/your/api/here').should.return200OnGet();
```

shraddhacloudelements replied:
As discussed, this does not work, so I am using `cloud.get()` in `it()`.

jjwyse commented on src/test/elements/adobe-esign/libraryDocuments.js:
```diff
@@ -0,0 +1,68 @@
+  });
+  it(`should allow GET ${test.api}/{libraryDocumentId}/documents/{documentId}`, () => {
+    let documentId;
+    sleep.sleep(30)
```
is it necessary to sleep for 30 seconds before the tests starts, and 30 seconds in the middle of the test?  just going to take a *long* time to run these tests if so....

shraddhacloudelements replied:
I am putting it to sleep for 60 seconds in the end of `before()` function, so that we can get rid of all other sleep() functions.

jjwyse commented on src/test/elements/adobe-esign/widgets.js:
```diff
@@ -0,0 +1,79 @@
+
+suite.forElement('esignature', 'widgets', null, (test) => {
+  /*
+  // This script breaks, as there are over 700 widgets which leads to time-out
```
don't want any tests that will break in here, since we rely on all these tests passing for many things. should we comment this out for now?

shraddhacloudelements replied:
It is already commented. Do you mean something else?

jjwyse commented on src/test/elements/adobe-esign/widgets.js:
```diff
@@ -0,0 +1,79 @@
+  });
+*/
+  it(`should allow GET ${test.api}/{widgetId}`, () => {
+    return cloud.get(test.api + '/' + widgetId);
```
same as comment above about using `test.withApi('/your/api/here').should.return200OnGet` here.

shraddhacloudelements replied:
As discussed, this does not work, so I am using `cloud.get()` in `it()`.

shraddhacloudelements commented:
## Notes
* Changed the name of the resources: agreementAssetEvents, libraryDocuments, transientDocument as per the CE APIs standard.
* Renamed the JS files as well and changed the scripts accordingly.
* Made all changes for the strings to be in ES6 literals format.
* After discussion with the team, to avoid all the POSTs in `library-documents` and `widgets`, created new Widgets Library Documents in the Adobe ESign UI and hardcoded their `widgetId` and `libraryDocumentId` in the `widgets.js` and`library-documents.js` respectively. Below are the details of them:

```
Widget details::
"name": "DoNotDeleteThisWidgetThisIsForChurrosTesting"
"widgetId": "3AAABLblqZhA_ZRfpH6d0l0tD8pnUXG-UWi0Xe0gtQ8dRJCvFzbQyMxiEyzUXcUWIi-xrslmcBBTlqU0DOukS_31TmSaqCck9"

Library Document details::
"name": "DoNotDeleteThisLibraryDocumentThisIsForChurrosTesting"
"libraryDocumentId": "3AAABLblqZhClHK1fioPebGw8EMx-PrHOTwkxSZMn6hfb0y3T95CA9ScNV7XZytrJM2gHPHMR0DgdY3simUO62FIYJnetl25d"
```
* Commented out the POST /users, to avoid creation of new users as there is no DELETE API for it.

jjwyse commented:
@shraddhacloudelements - looks good.  last thing: can you please add unit tests around the new functions you've added in `cloud.js`.  you'll notice from the `coveralls` comments above how it mentions "coverage has decreased" and that's because of no unit tests around those new functions.  besides that, looks good.

shraddha-jain commented:
## Notes

Added unit test for `crs` function in `cloud.test.js` and `suite.test.js`. Below are the test results:

```
  123 passing (2m)

=============================================================================
Writing coverage object [/Users/shraddhajain/Workspaces/churros/coverage/coverage.json]
Writing coverage reports at [/Users/shraddhajain/Workspaces/churros/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 97.77% ( 571/584 )
Branches     : 86.99% ( 127/146 )
Functions    : 100% ( 0/0 )
Lines        : 99.61% ( 515/517 )
================================================================================
```
